wallet: encapsulate transactions state

ariard opened the pull request:
While working on #15931, I've tried to rationalize tx state management to ease integration of block height tracking per-wallet tx. We currently rely on a combination of `hashBlock` and `nIndex` with magic value to determine tx confirmation, conflicted or abandoned state. It's hard to reason and error-prone.  To solve that, we encapsulate these fields in a `TxConfirmation` struct and introduce a `TxState` member that we update accordingly at block connection/disconnection.

Following jnewbery [recommendation](https://github.com/bitcoin/bitcoin/pull/15931#discussion_r312576506), I've taken these changes in its own commit, and open a PR to get them first. It would ease review of aforementioned PR, but above all should ease fixing of long-term issues like :
* https://github.com/bitcoin/bitcoin/issues/7315 (but maybe we should abandon abandontransaction or relieve it to only free outpoints not track the transaction as abandoned in itself, need its own discussion)
* https://github.com/bitcoin/bitcoin/issues/8692 where we should cancel conflicted state of transactions chain smoothly
* `MarkConflicted` in `LoadToWallet` is likely useless if we track conflicts rights at block connection

Main changes of this PR to get right are tx update in `AddToWallet` and serialization/deserialization logic.

Commit:
Encapsulate tx status in a Confirmation struct

Instead of relying on combination of hashBlock and nIndex
values to manage tx in its lifecycle, we introduce 4
status : CONFIRMED, UNCONFIRMED, CONFLICTED, ABANDONED.

hashBlock and nIndex magic values should only be used at
serialization/deserialization for backward-compatibility.

At block disconnection, we know flag txn as UNCONFIRMED where
previously they kept their states until being override by a
block connection or abandontransaction call. This is a change
in behavior for which user may have to call abandon twice
if transaction is disconnected and not accepted back in the mempool.

We assert status transitioning right in AddToWallet. Doing so
flagged a misbehavior in ComputeTimeSmart unit test where same
tx is confirmed twice in different block. To avoid inconsistencies
we unconfirmed tx before new connection in different block. We
also remove a cs_main lock in test, as AddToWallet and its
callees don't rely on locked chain.

Commit:
Remove SyncTransaction for conflicted txn in CWallet::BlockConnected

We shouldn't rely on this sync call to get an accurate view of txn
state, if a tx conflicts with one in mapTx we are going to update
our wallet dependencies in AddToWalletIfInvolvingMe while conflicting
txn get connected. If it doesn't conflict with one of our dependencies
we are not going to track it anyway.

This is a cleanup, as this SyncTransaction is redundant with the
following one for confirmation which is triggering the MarkConflicted
logic. We keep the loop because set of conflicted txn isn't same as txn
included in block.

ariard commented:
Travis is stalling yet another time..

jnewbery commented:
Concept ACK

ryanofsky commented:
Concept ACK. This overlaps (but is compatible) with #9381

Sjors reviewed (commented):
Concept ACK. I added tests for backwards compatibility of abandoned and conflicted wallet transactions to #12134. Those still pass if I apply this PR, which gives some assurance we didn't break the serialization.

practicalswift commented on src/wallet/wallet.h:
```diff
@@ -607,10 +651,14 @@ class CWalletTx
+    void setConflicted() { tx_state.state = CWalletTx::CONFLICTED; }
+    bool isUnconfirmed() const { return tx_state.state == CWalletTx::UNCONFIRMED; }
+    void setUnconfirmed() { tx_state.state = CWalletTx::UNCONFIRMED; }
+    bool isConfirmed() const { return tx_state.state == CWalletTx::CONFIRMED; }
```
`isConfirmed()` does not seem to be tested as the others. Add a test? :-)

ariard replied:
Removed for now, I've added it for further changes were ABANDON state was dual-set with other states, but drop if for now, would have been too big changes at once.

practicalswift commented on src/wallet/wallet.h:
```diff
@@ -477,16 +475,46 @@ class CWalletTx
+    /* New transactions start in the UNCONFIRMED. At BlockConnected,
+     * they will transition to CONFIRMED. In case of reorg, at BlockDisconnected,
+     * they rollback to UNCONFIRMED. If we detect a conflicting transaction at
+     * block connection, we update conflited tx and its dependencies as CONFLICTED.
```
Should be "conflicted"? :-)

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -477,16 +475,46 @@ class CWalletTx
+         * Older clients interpret nIndex == -1 as unconfirmed for backward
+         * compatibility (pre-commit 9ac63d6).
+         */
+        int nIndex;
```
> Can we move this backwards compatibility feature entirely to the serialization code, and then make `nIndex` a `uint` (for clarity)?

Personally -0 on this. Imo, unsigned types provide no safety and are terrible for everything except bitfields.

ariard replied:
What do you mean here by moving this backwards compatibility feature entirely to the serialization code ? If changes are right this logic should only lived in serialization code now.

Sjors replied:
See below, it depends on if nIndex has any meaning.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -477,16 +475,46 @@ class CWalletTx
+         * Older clients interpret nIndex == -1 as unconfirmed for backward
+         * compatibility (pre-commit 9ac63d6).
+         */
+        int nIndex;
```
Just a suggestion, but I'd add `= 0` here and `= UNCONFIRMED` on the line below to simplify constructor and deserialization code.

ariard replied:
done

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -477,16 +475,46 @@ class CWalletTx
         nChangeCached = 0;
         nOrderPos = -1;
+        tx_state.hashBlock = uint256();
+        tx_state.nIndex = 0;
```
> Why is the default initialization changed to 0?

From what I can tell in this PR, the convention mostly adhered to in this PR is to stop using -1 and ABANDON_HASH values at runtime, and restrict them only to the serialization code. This seems like a pretty good convention if it's followed consistently. I also think it'd be nice to explicitly say what values the block and index values should have for different states in a comment.

ariard replied:
Yes see comment on top of `TxConfirmation`, should be improved ?

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -513,7 +541,23 @@ class CWalletTx
+            tx_state.nIndex = 0;
+            setAbandoned();
+        } else if (serializedIndex == -1) {
+            tx_state.nIndex = -1;
```
Can we set 0 here, or change MarkConflicted to set -1? I think it'd be good to be consistent about what conflicted transactions look like.

ariard replied:
You're right that an inconsistency from my part. Corrected.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1415,7 +1415,6 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
     // the notification that the conflicted transaction was evicted.
 
     for (const CTransactionRef& ptx : vtxConflicted) {
-        SyncTransaction(ptx, CWalletTx::TxState::CONFLICTED, {} /* block hash */, 0 /* position in block */);
```
In commit "Remove SyncTransaction for conflicted txn in CWallet::BlockConnected" (0e3212ae85c7cf1eb6a8d937ac67d6822f624175)

I don't think I understand this change. Commit description suggests why it is ok, but I don't understand what motivates it. Is it just cleanup? Is it maybe a minor bugfix, or needed for a future change?

ariard replied:
Updated commit message. I think this redundant, as conflicts tagging logic is triggered by connection of conflicting transaction. AFAIK, it's not a bugfix but a cleanup. And as it was same issue than major commit, I bundled them together.

ariard replied:
Extended commit message, we keep the loop because set of conflicted txn isn't same as txn included in a block.

ryanofsky reviewed (commented):
Some questions, but this looks very good and is a welcome change that should make code more readable.

ariard commented on src/wallet/wallet.h:
```diff
@@ -477,16 +475,46 @@ class CWalletTx
+        uint256 hashBlock;
+        /* At serialization/deserialization, an nIndex == -1 means that hashBlock refers to
+         * the earliest block in the chain we know this or any in-wallet dependency conflicts
+         * with, but if hashBlock is null it means transaction is abandoned. An nIndex >= 0
```
With these changes, but even before, `nIndex` never referred directly to a block height. When set at -1, its interpreted as a magic value meaning the `hashBlock` is one of the deepest conflicting tx. This PR intents to scope this logic only to serialization/deserialization.

Sjors replied:
If `nIndex` doesn't mean a block height, then why not remove the instance variable altogether? We can still (de)serialize in a backwards compatible way.

ariard replied:
Well I think we need to keep `nIndex` at least it's used by RPCs in `WalletTxToJSON` ? Or getting it out of new `TxConfirmation` struct ?

Sjors replied:
I think we should figure out what it means in the RPC, consider deprecating that field, and then having an explicit backwards compatible workaround.

ariard commented on src/wallet/wallet.h:
```diff
@@ -477,16 +475,46 @@ class CWalletTx
+        uint256 hashBlock;
+        /* At serialization/deserialization, an nIndex == -1 means that hashBlock refers to
+         * the earliest block in the chain we know this or any in-wallet dependency conflicts
+         * with, but if hashBlock is null it means transaction is abandoned. An nIndex >= 0
```
Currently, `AbandonTransaction` set `nIndex` as -1 and this value is kept at serialization/deserialization. I'm not sure if it make sense for older client, because an abandoned transaction can be also confirmed/unconfirmed/conflicted at same time. Nevertheless, I've tried to keep same behavior for now, so `hashBlock` set as ABANDON_HASH is used to avoid deserialization ambiguity with conflicted.

I've clarified this comment, the ABANDON_HASH one and also reset abandon txn `hashBlock` at deserialization. ABANDON_HASH should be scoped at serialization/deserialization.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1249,8 +1238,7 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const uint256
             // Get merkle branch if transaction was found in a block
-            if (!block_hash.IsNull())
-                wtx.SetMerkleBranch(block_hash, posInBlock);
+            wtx.SetMerkleBranch(state, block_hash, posInBlock);
```
Clarified in commit message that is a change in behavior where at block disconnection, previous tx state is override by UNCONFIRMED one. I think right now, we don't track at all the fact that tx has been disconnected, and that's why we don't undo conflicts, or make it hard to solve them. The only direct consequence of these changes I can think of is a user having to call abandontransaction a 2nd time if block get disconnected and transaction is not back in mempool.

This change would make easier to track conflicts, where if conflicting transaction get UNCONFIRMED, iterate on conflicted outpoints and make them free.

Maybe, I can avoid update right now for UNCONFIRMED state, so a tx is born in this state but never go backward to it, and save implications for a later PR ?

ariard commented:
Thanks all for reviews, updated with comments corrected at ba86976.

meshcollider commented:
Concept ACK

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1383,8 +1372,9 @@ void CWallet::MarkConflicted(const uint256& hashBlock, const uint256& hashTx)
 
-void CWallet::SyncTransaction(const CTransactionRef& ptx, const uint256& block_hash, int posInBlock, bool update_tx) {
-    if (!AddToWalletIfInvolvingMe(ptx, block_hash, posInBlock, update_tx))
+void CWallet::SyncTransaction(const CTransactionRef& ptx, CWalletTx::TxState state, const uint256& block_hash, int posInBlock, bool update_tx)
```
`posInBlock` could be a good rename candidate for `nIndex`, though probably not worth making the diff bigger.

Sjors reviewed (approved):
ACK 6cc34fc

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -4628,21 +4617,24 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
 {
     // Update the tx's hashBlock
-    hashBlock = block_hash;
+    tx_state.hashBlock = block_hash;
```
Oh I've struggled hard on naming, `TxConfirmation` is struct with tx state + its values (block hash, position in block). If you have any better names, I'll take it :)

Sjors replied:
The variable name `tx_state` is the confusing part, because it's not a TxState

ariard replied:
Replaced by `tx_conf`

ariard commented on src/wallet/wallet.h:
```diff
@@ -477,16 +477,44 @@ class CWalletTx
-    int nIndex;
+    struct TxConfirmation {
+        uint256 hashBlock;
+        /* At serialization/deserialization, an nIndex == -1 means that hashBlock refers to
```
Moved to the deserialization code!

ariard commented:
Thanks @Sjors for review, pushed some corrections on 3e027c7

ryanofsky reviewed (approved):
utACK 3e027c776556c3070a9c3460045a6e397b21ceb8

ryanofsky reviewed (approved):
utACK f9f492638a488149acceb7b6487553862e919028. Only change since last review is renaming tx_state to tx_conf, which is reasonable. I left some comments with my previous review, but they aren't very important and are fine to ignore.

meshcollider reviewed (commented):
Looks good to me, utACK f9f492638a488149acceb7b6487553862e919028

>@ryanofsky My concern is that the new behavior might be less robust than previous behavior if there is a bug in higher level code that marks a transaction as confirmed in two different blocks without marking the transaction as unconfirmed in between.

I agree, I was looking at this specific change too. I think it should be fine but the additional check wouldn't hurt :+1:

ariard commented on src/wallet/wallet.h:
```diff
@@ -477,16 +477,37 @@ class CWalletTx
         fInMempool = false;
         nChangeCached = 0;
         nOrderPos = -1;
+        tx_state.hashBlock = uint256();
```
Well strictly speaking, given that Init set to unconfirmed, it would be more wrongly updating to confirmed. But if hash is set IMO we have database corruption or bug in higher tracking logic. Updated serialization logic with this and beneath to make it more robust.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1118,21 +1118,10 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
+            wtx.tx_state.nIndex = wtxIn.tx_state.nIndex;
+            wtx.tx_state.hashBlock = wtxIn.tx_state.hashBlock;
             fUpdated = true;
         }
```
Done

jnewbery replied:
(re: https://github.com/bitcoin/bitcoin/pull/16624#discussion_r315293956)

I don't like these new asserts. I think they could be triggered if:

- a wallet tx is confirmed in a block
- the wallet is shut down
- the block chain re-orgs and the tx is included in a different block
- the wallet file is loaded on a sync'ed node

All this code is trying to do is update an existing wallet tx's `Confirmation` status with the new `Confirmation` status. I think we can just change it to:

`if (wtxIn.tx_state.state != wtx.tx_state.state || wtxIn.tx_state.hashBlock != wtx.tx_state.hashBlock || wtxIn.tx_state.nIndex != wtx.tx_state.nIndex) {`

ariard replied:
If these new asserts get triggered it would mean we have bugs in higher level code, but yes aborting maybe too strong given that this kind of bugs could be due to some likely API misuse in RPCs ? What's about logging them with a warning ?

I'm not in favor of proposed alternative as it may cause hidden inconsistencies if this aforementioned high level bugs are present

ryanofsky replied:
I don't feel strongly, but I think asserts are way to go because we should not allow bugs and uncertainty in wallet event processing code. It's hard for me to imagine a scenario where it would be a win to use defensive programming in a place like this.

jnewbery replied:
@ryanofsky - do you agree that the scenario I describe above would trigger the assert?

ryanofsky replied:
Oh, I didn't realize https://github.com/bitcoin/bitcoin/pull/16624#discussion_r315883326 was describing a specific scenario, not just a list of possible events. I could see that scenerio triggering these asserts, and making the change you suggested an improvement and bugfix. So it is likely some change is needed here, and the suggested one seems ok.

More ideally, though, it'd be really nice to have a functional test that triggers these asserts, and instead of hiding the problem of an unconfirmed transaction being left in the CONFIRMED state after a reorg, there would be code on startup that marks reorged transactions UNCONFIRMED before rescanning (could basically be the same code from #15931 that looks up block heights on startup).

Nice catch though (assuming this a bug)!

jnewbery replied:
> I didn't realize #16624 (comment) was describing a specific scenario

sorry - changed my unordered list to an ordered list and updated wording to make this clearer.

jnewbery replied:
> there would be code on startup that marks reorged transactions UNCONFIRMED before rescanning

Yes - I think if the wallet rescans from block at height x, it should first change all transactions that are CONFIRMED at height >x to UNCOFIRMED.

ariard replied:
I let asserts and move some bits from #15931 to transition status to UNCONFIRMED if blockHash not anymore in chain. Also added a wallet_reorgsrestore test with aforementioned scenario, without new code it triggers asserts

ariard commented:
Updated at 5ef9e95 with @ryanofsky comments. Main changes are a more robust serialization logic and assert for already present txn in AddToWallet. This last one flagged a misbehavior of AddToWallet in ComputeTimeSmart (see commit message), so was worth adding.

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -590,7 +629,7 @@ class CWalletTx
     std::set<uint256> GetConflicts() const NO_THREAD_SAFETY_ANALYSIS;
 
-    void SetMerkleBranch(const uint256& block_hash, int posInBlock);
+    void SetMerkleBranch(Status status, const uint256& block_hash, int posInBlock);
```
nit: this function is badly named (and has been since 391dff16fe9ace90fc0f3308a5c63c453370e713 _Do not store Merkle branches in the wallet._). Now would be a good time to rename it to `SetTxConf()` or similar.

jnewbery commented:
Just nits so far. Still planning on doing a more thorough review.

ariard commented on src/wallet/wallet.h:
```diff
@@ -477,16 +477,37 @@ class CWalletTx
+
+    /* TxConfirmation includes tx state and a pair of block hash/block index at which tx has been confirmed.
+     * This pair is both 0 if tx hasn't confirmed yet. Meaning of these fields changes with CONFLICTED state
+     * where they instead point to block hash and index of the deepest conflicting tx.
```
Yes that's right, I rewrote comment to new code without thinking it wasn't relevant anymore

jnewbery replied:
Oh, I think you've removed too much now. It's still useful to know that `hashBlock` refers to the block hash of the deepest conflicting transaction.

ariard commented on src/wallet/wallet.h:
```diff
@@ -477,16 +477,37 @@ class CWalletTx
+        TxState state = UNCONFIRMED;
+    };
+
+    TxConfirmation tx_conf;
```
Yes already corrected at 5ef9e95

ariard commented on src/wallet/wallet.h:
```diff
@@ -513,7 +534,27 @@ class CWalletTx
+        s >> tx >> tx_conf.hashBlock >> dummy_vector1 >> serializedIndex >> dummy_vector2 >> mapValue >> vOrderForm >> fTimeReceivedIsTxTime >> nTimeReceived >> fFromMe >> dummy_char;
+
+        /* At serialization/deserialization, an nIndex == -1 means that hashBlock refers to
+         * the earliest block in the chain we know this or any in-wallet dependency conflicts
```
Updated, just to be sure a 'in-wallet dependency' is relative to a parent tx on which it relies to be valid ?

ariard commented:
Rebased baaf604 with @jnewbery nits and comments addressed. Thanks for the review!

Still need to decide on https://github.com/bitcoin/bitcoin/pull/16624#discussion_r316231071 which spotted this https://github.com/bitcoin/bitcoin/pull/16624#issuecomment-523150156

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -477,16 +477,36 @@ class CWalletTx
-    int nIndex;
+    struct Confirmation {
+        Status status = UNCONFIRMED;
+        uint256 hashBlock = uint256();
```
In commit "Encapsulate tx status in a Confirmation struct" (e591ef7a796d1ca164e791862e20a31b801fdf1b)

Maybe add a doxygen comment for hashBlock saying it's set to the block of the deepest conflicting transaction when the wallet transaction is in the CONFLICTED state. This information seems to have been lost in latest change to the comments above.

ariard replied:
Reestablished

ryanofsky reviewed (approved):
utACK 63596cf1c56e1adb000f95cfce7aba5d96e838c2. Left some minor suggestions, feel free to ignore. Changes since last review: various field/method renames, tweaking setter methods to update multiple fields at once, comment updates, removing "TODO: Temporarily ensure" comment that I never understood, resetting all fields instead of just enum field in Init(), so the nIndex field is now reset during deserialization when transaction is unconfirmed.

Commit:
Modify wallet tx status if has been reorged out

Add a LockChain method to CWallet to know if we can lock or query
chain state safely.

At tx loading, we rely on chain to know if hashBlock of tx is still
in main chain. If not, we set its status to unconfirmed and reset
its hashBlock/nIndex.

If wallet loaded is the wallet-tool one, all wallet txn will
show up with a height of zero. It doesn't matter as status is not
used by wallet-tool.

We take lock prematurely in CWallet::LoadWallet and CWallet::Verify
to ensure that lock order is respected between cs_main an cs_wallet.

Commit:
Add a test wallet_reorgsrestore

Test we change tx status at loading in case of reorgs while wallet
was shutdown.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1248,9 +1240,9 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const uint256
-                wtx.SetMerkleBranch(block_hash, posInBlock);
+            // Block disconnection override an abandoned tx as unconfirmed
+            // which means user may have to call abandontransaction again
+            wtx.SetConf(status, block_hash, posInBlock);
```
Yes I've tried to combine abandon state with every other one but would run into issue. IMO, would be better to rework abandontransaction to something like abandoncoins, because confirmation are on the transaction-level, but conflict/abandon at the output one

ariard commented:
Updated at 12fea46 with new commit and test to solve https://github.com/bitcoin/bitcoin/pull/16624#discussion_r315883326

Sjors commented:
Travis machines 5 and 8 will fail until the next rebase (see #16561).

MarcoFalke commented:
Travis should run the latest version and a rebase is not required.

Have you tried running the tests locally with `./configure --enable-debug`?

MarcoFalke commented:
```
./out/x86_64-unknown-linux-gnu/bin/test_bitcoin -t wallet_tests
Running 9 test cases...
Assertion failed: detected inconsistent lock order at sync.cpp:121, details in debug log.
unknown location(0): fatal error: in "wallet_tests/ComputeTimeSmart": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp(303): last checkpoint

MarcoFalke commented:
```


2019-08-23T17:36:07Z Bitcoin Core version v0.18.99.0-unk (debug build)
2019-08-23T17:36:07Z Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2019-08-23T17:36:07Z Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2019-08-23T17:36:07Z Opened LevelDB successfully
2019-08-23T17:36:07Z Using obfuscation key for /tmp/test_common_Bitcoin Core/1566581767_852108204/blocks/index: 0000000000000000
2019-08-23T17:36:07Z Opened LevelDB successfully
2019-08-23T17:36:07Z Wrote new obfuscate key for /tmp/test_common_Bitcoin Core/1566581767_852108204/chainstate: ab46c080088c71f8
2019-08-23T17:36:07Z Using obfuscation key for /tmp/test_common_Bitcoin Core/1566581767_852108204/chainstate: ab46c080088c71f8
2019-08-23T17:36:07Z Pre-allocating up to position 0x1000000 in blk00000.dat
2019-08-23T17:36:07Z UpdateTip: new best=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f height=0 version=0x00000001 log2_work=32.000022 tx=1 date='2009-01-03T18:15:05Z' progress=0.000000 cache=0.0MiB(0txo)
2019-08-23T17:36:07Z ERROR: DeserializeFileDB: Failed to open file /tmp/test_common_Bitcoin Core/1566581767_852108204/banlist.dat
2019-08-23T17:36:07Z Invalid or missing banlist.dat; recreating
2019-08-23T17:36:07Z [default wallet] Wallet File Version = 10500
2019-08-23T17:36:07Z [default wallet] Keys: 0 plaintext, 0 encrypted, 0 w/ metadata, 0 total. Unknown wallet records: 0
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z) POTENTIAL DEADLOCK DETECTED
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z) Previous lock order was:
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z)  (1) wallet.cs_wallet wallet/test/wallet_tests.cpp:283 (in thread )
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z)  (1) cs_wallet wallet/wallet.cpp:1091 (in thread )
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z)  (2) cs_main interfaces/chain.cpp:255 (in thread )
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z) Current lock order is:
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z)  (2) cs_main interfaces/chain.cpp:245 (in thread )
2019-08-23T17:36:07Z (mocktime: 1970-01-01T00:01:40Z)  (1) cs_wallet wallet/wallet.cpp:3352 (in thread )

ariard commented:
Updated at be2d6d3 with a lock order correction. Thanks to @MarcoFalke for spotting #16700 and finding the root issue with failures

ryanofsky reviewed (approved):
utACK be2d6d3cd1421d66871396eeef5f6e30b0841889. Only changes since last review are adding comments and adding back a removed test cs_main lock in the first commit, and adding third and fourth commits addressing reorg during shutdown bug https://github.com/bitcoin/bitcoin/pull/16624#discussion_r315883326

I think locking changes in the third commit can be simplified and I left some review comments, but feel free to ignore them. I think the current approach is fine given that #16426 will get rid of this locking pretty much altogether.

ariard commented:
Updated at 442a87c with some @ryanofsky comments to take lock smoother. 
Note, also fix a race condition in the new wallet_reorgsrestore, were order of block announcements after block connection was making test successful or not.

ryanofsky reviewed (approved):
utACK 442a87cc0ae43ebc9b6654a6165778eecb931f74. Changes since last review are switching from `hasChain` to `LockChain` and removing chain lock in `WalletBatch::LoadWallet` that's redundant with the new lock still added in `CWallet::LoadWallet`, and fixing python test race condition.

meshcollider commented:
Light re-Code Review ACK 442a87cc0ae43ebc9b6654a6165778eecb931f74

I think this is in a good state, thank you @ryanofsky for being so active with review and @ariard for addressing feedback so quickly :)

jnewbery reviewed (commented):
ACK 442a87cc0ae43ebc9b6654a6165778eecb931f74 with a couple of nits inline.

Thanks for adding the test! An alternative approach to connecting/disconnecting and stop/starting the nodes would have been to use the `loadwallet`/`unloadwallet` and `invalidateblock`/`reconsiderblock` RPCs. I think that may have been clearer and faster to run.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1425,11 +1419,11 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
 
     for (const CTransactionRef& ptx : vtxConflicted) {
-        SyncTransaction(ptx, {} /* block hash */, 0 /* position in block */);
+        SyncTransaction(ptx, CWalletTx::Status::CONFLICTED, {} /* block hash */, 0 /* position in block */);
```
In commit "Encapsulate tx status in a Confirmation struct" (a31be09bfd77eed497a8e251d31358e16e2f2eb1)

This line is removed in the immediate next commit (https://github.com/bitcoin/bitcoin/pull/16624/commits/7e89994133725125dddbfa8d45484e3b9ed51c6e), but technically it has some bugs. Original intent of this code was only ever to send `-walletnotify`notifications, not to update transaction state, and the change should have used `UNCONFIRMED` instead of `CONFLICTED` to stay equivalent to previous code. Problems with the change:

- It was an undocumented change in behavior in what was supposed to be a refactoring commit
- It forgot to pass the block hash to `SyncTransaction` so wasn't properly setting a conflicted state. Instead it set an inconsistent state that would sometimes be treated as unconfirmed, sometimes conflicted.
- It would compound reorg problems wallet already has treating transactions that are no longer conflicted as conflicted, see [*bug-stale-conflicted*](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#bug-stale-conflicted).

In the future it would actually be possible to mark these transactions as conflicted instead of unconfirmed. Way this could be done is described [*idea-more-conflicts*](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#idea-more-conflicts).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1440,8 +1434,12 @@ void CWallet::BlockDisconnected(const CBlock& block) {
     auto locked_chain = chain().lock();
     LOCK(cs_wallet);
 
+    // At block disconnection, this will change an abandoned transaction to
+    // be unconfirmed, whether or not the transaction is added back to the mempool.
+    // User may have to call abandontransaction again. It may be addressed in the
+    // future with a stickier abandoned state or even removing abandontransaction call.
```
In commit "Encapsulate tx status in a Confirmation struct" (a31be09bfd77eed497a8e251d31358e16e2f2eb1)

Comment seems slightly off. I think the comment is right that `abandontransaction` would need to be called again in this case, but I don't think it would just be because of this code. The abandoned state would already have been lost previously when the block was connected.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1410,22 +1410,14 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
-    // the notification that the conflicted transaction was evicted.
 
-    for (const CTransactionRef& ptx : vtxConflicted) {
-        SyncTransaction(ptx, CWalletTx::Status::CONFLICTED, {} /* block hash */, 0 /* position in block */);
```
In commit "Remove SyncTransaction for conflicted txn in CWallet::BlockConnected" (7e89994133725125dddbfa8d45484e3b9ed51c6e)

Dropping this SyncTransaction call was a bug, since it caused `-walletnotify` notifications to be lost. These were restored later in #18982, basically just reverting this commit, but with some updates due to notification code changes in the meantime, and using `CONFIRMED` instead of `CONFLICTED` to avoid a bug from the previous commit https://github.com/bitcoin/bitcoin/pull/16624/commits/a31be09bfd77eed497a8e251d31358e16e2f2eb1

ryanofsky reviewed (commented):
Note: After this PR was merged, some unexpected behavior changes were found. One bug caused, another one partially fixed. See 16624 in the history section this wiki page https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#history for details
